Add users table

LordMilton opened the pull request:
Add Users table and basic queries for adding/editing users

LordMilton committed:
Implement /user/login for adding users to user table on auth0 login

Users are tracked as a username (email) and a user_id (assigned by auth0
as the sub claim on the jwt). Reviews are not yet verified against
actual, existing user_ids.

LordMilton commented:
Not tested, would prefer to merge https://github.com/latenitekid/food-review/pull/3 first to make testing less annoying

LordMilton committed:
Make all previous scripting runnable from main.py

Previous scripts (i.e. serve and migrations) are now run through main.py
where serve is default and migrations is run when -m is used as a cmd
line arg.

Also adds --production as a cmd line arg, but it's not used yet

Migrations was previously broken because running it as its own project
meant it couldn't actually access the config_utils. Running it this way
fixes that issue.

LordMilton commented on src/backend/migrations/20250905_1404_add_users_table.sql:
```diff
@@ -0,0 +1,10 @@
+-- Create reviews table
+CREATE TABLE IF NOT EXISTS users (
+    user_id VARCHAR(30) PRIMARY KEY,
```
I just realized, I went with the choice of using the 'sub' claim from the jwt, but I think it's hexadecimal so I just stored it as a string. I could either assume its hexadecimal and handle all that as an int, otherwise we'd have to handle all our user_id's as strings where user_id is a foreign key (requires front-end changes)

Let me look into the former first

Edit: jwt sub claim is only unique with the prepended 'auth0' or whatever else it is depending on login method and is therefore not hexadecimal. Need to interpret is a string

LordMilton replied:
_sigh_
So auth0 claims that the email identifier on the jwt is unique, it is **not** because you can make a login with a gmail and then separately make a login via google under the same gmail. I think the strategy will be to track user email as the username, the 'sub' as the unique identifier coming from auth0, and have a serial user_id to use as our foreign keys.

latenitekid replied:
Okie dokie, that sounds good to me

latenitekid commented on src/backend/migrations/20250905_1404_add_users_table.sql:
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE IF NOT EXISTS users (
+    user_id VARCHAR(30) PRIMARY KEY,
+    username VARCHAR(255) NOT NULL,
+    created_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_DATE
```
Should we use CURRENT_TIMESTAMP instead of DATE? Can't think of a specific reason we'd need the exact time, but couldn't hurt.

latenitekid commented on src/backend/security/auth0.py:
```diff
@@ -0,0 +1,31 @@
+                          algorithms=["RS256"],
+                          key=signing_key.key,
+                          options={"verify_aud": False}, # Audience isn't verifying successfully, front-end not putting correct audience on the jwt or am I?
+                          audience="http://localhost:8000")
```
Needs to match exactly, including the forward slash at the end:

`http://localhost:8000/`

I tried it with that small change and flipping verify_aud to True and it worked.

latenitekid replied:
And on that note, minor suggestion would be that it'd be nice for it to come from either a config or environment variable.

LordMilton replied:
Ah, thank you for solving that. I'll look into handling that in the config

latenitekid reviewed (commented):
The new arg parser is cool. I tested it as-is (though I know you plan to change schema a bit) and it works well.

LordMilton committed:
Use auth sub claim and user_id for user identification

Auth sub claim identifies users uniquely coming from auth0 (email does
not fulfill the uniqueness requirement). user_id is
for use as a foreign key.

LordMilton committed:
Fix auth audience and config for backend server

Backend server location is now dictated by configs. Auth0 stuff is now
verifying jwt audience.

LordMilton committed:
Clarify necessary config fields
